File: Array/strStr.py
## Implement strStr().
## Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

class Solution:
    # @param haystack, a string
    # @param needle, a string
    # @return an integer
    def strStr(self, haystack, needle):
        m, n = len(haystack), len(needle)
        if(not m): return -1 if n else 0
        if(not n): return 0
        for i in range(m):
            for j in range(n):
                if(haystack[i + j] != needle[j]): break
                if(j == (n - 1)): return i + 1
                if((i + j) == (m - 1)): return -1 
        return -1

File: Array/strStr.java
public class Solution {

	/** O(nm) runtime, O(1) space – Brute force: */ 
	
	public int strStr(String haystack, String needle){
		
		int m = needle.length(); 
		int n = haystack.length(); 
		if( m == 0) return 0; 
		if( m > n || (n == 0 && m != 0)) return -1; 
		
		int index; 
		for(int i = 0; i < n ; i++){
			int j = 0; 
			index = i; 
			// check length of remaining haystack: if enough to hold needle
			if(n - index >= m ){
				while(haystack.charAt(index) == needle.charAt(j) && j < m ){
					if(j + 1 == m) return i; 
					index ++; j ++;
				}
			}
			
		}
		return -1 ; 

	}
}
